Dataproduct publishing updated with dynamically added input fields and styled according to Marble design

alexandercyu opened the pull request:
- form styled according to Marble design
- dynamically added input field sections for Author and Geometry updated 
  -> first row has a remove button that is dynamically shown and hidden
  -> each row has a remove button to remove that particular row
  -> rows can be removed in any order
  -> indexes for rows generated from last remaining / first row (indexes may not necessarily increment from '1')
  -> Geometry section inputs can be swapped to display different geometry inputs without erasing previously entered information
  -> IDs of input fields made distinguishable by a geometry prefix

alexandercyu commented on templates/partials/dropdown-publish.html:
```diff
@@ -0,0 +1,86 @@
+                {
+                    "item_label": "Point",
+                    "item_id": "geometryPoint",
+                    "item_value": "Point"
```
Left unintentionally.  Removed this entry for all dropdown items.

alexandercyu commented on templates/site/publish-dataproduct.html:
```diff
@@ -0,0 +1,371 @@
+            }
+
+              /*
+            document.getElementById('geo_bbox').innerHTML = `
```
Added comments to say Cassie's input is needed before it can be deleted.

alexandercyu commented on static/css/form.css:
Removed the file and classes like 'child' and 'parent' from the html and js files.

mishaschwartz commented on templates/site/publish-dataproduct.html:
```diff
@@ -0,0 +1,375 @@
+    <script src="js/multi-polygon.js"></script>
+    <script>
+
+    function geoPolygon2(selected_geometry) {
```
put this in a js file

alexandercyu replied:
Moved to multi-polygon.js

mishaschwartz commented on templates/site/publish-dataproduct.html:
```diff
@@ -0,0 +1,375 @@
+    function geoPolygon2(selected_geometry) {
+        //var selected_geometry = document.getElementById("geometry").selectedIndex;
+        //console.log(selected_geometry);
+        if (selected_geometry == 1) {
```
let's make all of these one switch statement and remove all the if/else to simplify code

alexandercyu replied:
Moved all html javascript into multi-polygon.js or deleted it because it was unneeded.

mishaschwartz commented on templates/site/publish-dataproduct.html:
```diff
@@ -0,0 +1,375 @@
+                swapDiv('geo_point');
+                initializePointInputDiv('point', 'geo_point');
+            }
+            // TODO: Ask Cassie if the IDs or anything else in the input fields are used in the form processing. Delete if unused
```
No IDs are not necessary to send to the backend. This can be removed.

alexandercyu replied:
Removed all old code and  commented code in publish-dataproduct.html that referenced form IDs.

mishaschwartz reviewed (commented):
TODO:

1. remove commented code
2. remove/comment code that processes form inputs (this needs to be rewritten anyway)
3. add validations to inputs (https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input)
    - email for email inputs
    - numeric for lat and lon (with appropriate min/max values)
    - geojson input should be textarea
    - link to inputs should be textarea (helper message saying 1 per line)
    - link to additional files should be textarea (helper message saying 1 per line)
    - ... anything that can take multiple values as a comma or newline separated list 
should be a text area
    - require at least 3 points for polygon
    - require at least 2 points for line-string
4. Form processing: 
    - one function called on submit that stores values that are in the form to an object
    - don't worry about sending it somewhere yet
    - remember to only store geo values that are visible to the user (store type of shape, and lat/lon values)

alexandercyu commented:
- Add processing of author input

alexandercyu commented:
- changed Geometry processing to handle case where user selects a geometry with no coordinate input and has to paste in a geojson
- geojson that is pasted in will have newline characters stripped
- NOTE: input type datetime-local doesn't work properly in Firefox - the time selector doesn't show.

alexandercyu commented:
Add function to check the number input for latitude and longitude.  Function will show a warning if value is outside of the range for latitude or longitude.

NOTE: Currently has errors: function only fires on first key press/input, not for every key press/input.

alexandercyu commented:
- changed calendar input type back to 'datetime-local'
- made calendar icons clickable and put focus on date input fields
- all textareas now use newline parsing
- all textarea helper text now use clickable info icon
- added function to add the 'required' attribute and label text to the minimum remaining Author and coordinate input fields when removing an input row for those inputs
- added style for invalid inputs

alexandercyu commented:
- added checkboxes for calendar input
  -> makes end date same as start date
  -> disables both date inputs for case where there is no temporal data

- reduced Geometry dropdown items by removing Multi Linestring, Multi Polygon, and GeoJSON Collection
- added Upload GeoJSON item to Geometry dropdown

alexandercyu commented:
- changed checkbox for copying Start Date to End Date so the Start Date input is still enabled after checking the checkbox
- invalid styles applied during validation process

alexandercyu commented on templates/site/publish-dataproduct.html:
```diff
@@ -0,0 +1,211 @@
+            <div id="calender_checkbox_container" class="calendar-checkbox-container">
+                {% with %}
+                    {% set checkbox_id = "date_make_equal" %}
+                    {% set direction_left = true %}
```
That is odd.  If the direction_left checkbox is inspected the HTML is rendered correctly.

Nevertheless, they should both be on the right side.

Fixed so both are direction_right.

alexandercyu commented on templates/site/js/dataproduct-form.js:
```diff
@@ -0,0 +1,904 @@
+
+            break;
+
+        case 8:
```
Fixed to 6.

alexandercyu commented on templates/partials/dropdown-publish.html:
```diff
@@ -0,0 +1,75 @@
+
+
+
+{% if dropdown_type == "default_dropdown_list" %}
```
Removed if-statement.

alexandercyu commented on templates/site/js/dataproduct-form.js:
```diff
@@ -0,0 +1,904 @@
+        var labelCoordinateIDPrefix = "label_" + geometryType;
+
+        inputArray = document.querySelectorAll(`[id^=${coordinateIDPrefix}]`);
+        switch (inputArray.length){
```
Moved duplicated code to a helper function.

alexandercyu commented:
Added comment to note behaviour of checkbox label from the template.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -0,0 +1,904 @@
+
+            if(geometryContainer.classList.contains("show")){
+                if(geometryFile.value.indexOf("\n") > -1){
+                    geometryGeoJSONBoundingBoxInput = geometryFile.value.replaceAll("\n", "");
```
This looks great. The only edge case I think you missed is if the textarea is empty, then the validation error message does not show and the textarea is not considered required.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -0,0 +1,904 @@
+    // Example of how polygon coordinates should be formatted
+    //  [ [ 19.732387792295356,17.362269487080525],[ 15.018600022717294, 11.450658457798042],[29.998879431116166, 6.631460284225298] ]
+    var submitObject = {};
+    var geojsonTemplate =     {
```
Looks good except remember that case matters for the value of `type` so it should be "Point" not "point", etc.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -863,18 +938,29 @@ function submitForm(){
+    if(otherMetadataInputFields.length > 0){
+        for (otherInput of otherMetadataInputFields) {
+            var otherInputIDArray = otherInput.id.split("_");
+            geometryType = otherInputIDArray[0];
```
why are we talking about geometries now?

alexandercyu replied:
This is just how my mindset was when I was making the form.  IDs for input fields changed into the format of "category_indexnum".  I just kept using geometry type as the category.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -863,18 +938,29 @@ function submitForm(){
+        for (otherInput of otherMetadataInputFields) {
+            var otherInputIDArray = otherInput.id.split("_");
+            geometryType = otherInputIDArray[0];
+            geometryContainer = document.getElementById(geometryType + "_box");
```
unused?

alexandercyu replied:
Removed the unused item.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -863,18 +938,29 @@ function submitForm(){
-        for(dateMetadata of dateMetadataFields){
-            metadataObject[dateMetadata.id] = dateMetadata.value;
+    /*Add Metadata Other input to metadataObject*/
+    if(otherMetadataInputFields.length > 0){
```
Won't this always be true since the form starts with at least one row?

alexandercyu replied:
Removed the condition.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -863,18 +938,29 @@ function submitForm(){
+            geometryType = otherInputIDArray[0];
+            geometryContainer = document.getElementById(geometryType + "_box");
+
+            if (otherInput.value != "") {
```
If a key with no value is invalid (and won't be sent to the backend) then it should be marked invalid on the form.
Note that you probably want to treat a row with no key and no value differently from a form with no key and a value

alexandercyu commented:
Added link to Essential Climate Variables for the Climatological Variables section.
